packages/eslint-plugin/src/rules/no-unnecessary-condition.ts

```diff
@@ -0,0 +1,144 @@
+import {
+  TSESTree,
+  AST_NODE_TYPES,
+} from '@typescript-eslint/experimental-utils';
+import ts, { TypeFlags } from 'typescript';
+import {
+  isTypeFlagSet,
+  unionTypeParts,
+  isFalsyType,
+  isBooleanLiteralType,
+  isLiteralType,
+} from 'tsutils';
+import {
+  createRule,
+  getParserServices,
+  getConstrainedTypeAtLocation,
+} from '../util';
+
+// Truthiness utilities
+// #region
+const isTruthyLiteral = (type: ts.Type): boolean =>
+  isBooleanLiteralType(type, true) || (isLiteralType(type) && !!type.value);
+
+const isPossiblyFalsy = (type: ts.Type): boolean =>
+  unionTypeParts(type)
+    // PossiblyFalsy flag includes literal values, so exclude ones that
+    // are definitely truthy
+    .filter(t => !isTruthyLiteral(t))
+    .some(type => isTypeFlagSet(type, ts.TypeFlags.PossiblyFalsy));
+
+const isPossiblyTruthy = (type: ts.Type): boolean =>
+  unionTypeParts(type).some(type => !isFalsyType(type));
+// #endregion
+
+type ExpressionWithTest =
+  | TSESTree.ConditionalExpression
+  | TSESTree.DoWhileStatement
+  | TSESTree.ForStatement
+  | TSESTree.IfStatement
+  | TSESTree.WhileStatement;
+
+export type Options = [
+  {
+    ignoreRhs?: boolean;
+  },
+];
+
+export default createRule<Options, 'alwaysTruthy' | 'alwaysFalsy' | 'never'>({
+  name: 'no-unnecessary-conditionals',
+  meta: {
+    type: 'suggestion',
+    docs: {
+      description:
+        'Prevents conditionals where the type is always truthy or always falsy',
+      category: 'Best Practices',
+      recommended: false,
+      requiresTypeChecking: true,
+    },
+    schema: [
+      {
+        type: 'object',
+        properties: {
+          ignoreRhs: {
+            type: 'boolean',
+          },
+        },
+        additionalProperties: false,
+      },
+    ],
+    messages: {
+      alwaysTruthy: 'Unnecessary conditional, value is always truthy.',
+      alwaysFalsy: 'Unnecessary conditional, value is always falsy.',
+      never: 'Unnecessary conditional, value is `never`',
+    },
+  },
+  defaultOptions: [
+    {
+      ignoreRhs: false,
+    },
+  ],
+  create(context, [{ ignoreRhs }]) {
+    const service = getParserServices(context);
+    const checker = service.program.getTypeChecker();
+
+    /**
+     * Checks if a conditional node is necessary:
+     * if the type of the node is always true or always false, it's not necessary.
+     */
+    function checkNode(node: TSESTree.Node): void {
+      const tsNode = service.esTreeNodeToTSNodeMap.get<ts.ExpressionStatement>(
+        node,
+      );
+      const type = getConstrainedTypeAtLocation(checker, tsNode);
+
+      // Conditional is always necessary if it involves `any` or `unknown`
+      if (isTypeFlagSet(type, TypeFlags.Any | TypeFlags.Unknown)) {
```

Comment:
do we want to add an option to disallow `any`?
Do you think that some people might appreciate a stricter option?

Reply:
As far as I can come up with, the only other way to handle `any` would be to assume that any condition involving `any` is unnecessary.  I don't think that would be helpful, as most of those linter errors would not actually be unnecessary condition issues.  That seems more suited to a rule aimed at preventing `any`-typed values.